doc/fpw.1
.\"  This program is free software; you can redistribute it and/or modify
.\"  it under the terms of the GNU General Public License as published by
.\"  the Free Software Foundation; either version 2 of the License, or
.\"  (at your option) any later version.
.\"
.\"  This program is distributed in the hope that it will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"  GNU General Public License for more details.
.\"
.\"  You should have received a copy of the GNU General Public License
.\"  along with this program; if not, write to the Free Software
.\"  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"

.TH FPW 1

.SH NAME
.B fpw
\- FootPrintWizard for the printed circuit board layout tool.

.SH SYNOPSIS
.B fpw
[--help | -h]
[--quiet |--silent | -q]
[--verbose | -v]
[--version | -V]
[--help | -h]
[--format <footprint wizardfilename> | -f <footprintwizard filename>]
[--output <footprint filename> | -o <footprint filename>]

.SH DESCRIPTION
The
.B fpw
program is a tool for the creation of footprints to be used in the layout of printed circuit boards.
The complete manual for
.B fpw
is provided in a GNU texinfo format as well as HTML and PDF.
The texinfo version of the manual is typically viewed with the
.B info
program or alternatively with
.B emacs
or a graphical info viewer such as
.B tkinfo
.  The PDF and HTML documentation is typically installed as
/usr/local/share/fpw/fpw.html and /usr/local/share/fpw/fpw.pdf.
The prefix "/usr/local" may vary at your site.

.SH OPTIONS
.l
\fIfpw\fP accepts the following options:
.TP 8
.B -h
.TP 8
.B --help
Print a help message with command line options. (optional)
.TP 8
.B -q
.TP 8
.B --quiet
.TP 8
.B --silent
.TP 8
Quiet mode on.  This mode turns off all warnings/notes/messages. (optional)
.TP 8
.B -v
.TP 8
.B --verbose
Verbose mode on.  This mode gives as much feedback to the user as possible. (optional)
.TP 8
.B -h
.TP 8
.B --help
Print a message explaining the usage of the command line version of the program. (optional)
.TP 8
.B -V
.TP 8
.B --version
Print the version of the program. (optional)
.TP 8
.B -f
.TP 8
.B --format <footprintwizard filename>
Specify a filename for the program to read the resources for the footprint from.
.TP 8
.B -o
.TP 8
.B --output <footprint filename>
Specify a filename for the footprint to be created.

.SH EXAMPLES
This example assumes that you have a footprintwizard file called BGA10X10.fpw in
the current directory.
.TP 8
.B fpw --quiet --format BGA10X10.fpw --output BGA50C10X10-100_1100X1100.fp
.TP 8
to create a footprint file while having no warnings, info or error log messages whatsoever.

.SH "ENVIRONMENT"
\fIfpw\fP respects the following environment variable:
.PP
.TP 8
.B <none>

.SH "AUTHOR"
Bert Timmerman and maybe others.

.SH SEE ALSO
.BR ... (1)

.SH COPYRIGHT
.nf
Copyright \(co  2007-2011 Bert Timmerman.

This document can be freely redistributed according to the terms of the
GNU General Public License version 2 or any later version.


ChangeLog
commit a0df3e7faaa0f739fa5b4ad01f732a2fd9180fc3
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Sat Mar 5 11:56:29 2011 +0100

    Udated the German translation file.

commit 71a3f8c33b89fbaef829a3bef09a84688a4870e8
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Sat Mar 5 10:28:30 2011 +0100

    Updated the Dutch translation file with minor corrections.

commit 3bba19a4eaa6997ab6474c92e8509eeb37b925bc
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Sat Mar 5 10:25:59 2011 +0100

    Updated the translation template file after minor corrections.

commit c018909c47daf7e10450dbc3527622a4f2dd7ecf
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Fri Mar 4 07:47:58 2011 +0100

    BGA: correction of a typo in an error message.

commit 0738abc98cc519234c86387b3bbc16ca692e7315
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Wed Mar 2 23:13:12 2011 +0100

    Corrected the typo in successfull.

commit 91b019d2ec1e2842954158542e142f62b994f02c
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Wed Mar 2 22:19:08 2011 +0100

    Correction of a typo in an error message.

commit 7623f2f02fb7d14067a2714074082cebe24a2b77
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Tue Mar 1 17:30:06 2011 +0100

    Updated the Dutch translation file.

commit 0206b9d2b13e04408800ecc2910ef179ae73e4e7
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Tue Mar 1 17:29:31 2011 +0100

    Updated the translation template file.

commit 20ae54518a42fdd8e45c3ed8d705b2865739fc2c
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Tue Mar 1 17:28:54 2011 +0100

    fpw: updated the version message.

commit 1035d6985490233feb45ee16a744d1d06d0eaa58
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Mon Feb 28 23:24:50 2011 +0100

    libfpw: include string.h to satisfy memset and strtok.

commit 23f6a9e175b7d14cf8160fcc335daeecc714eac0
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Mon Feb 28 23:19:09 2011 +0100

    Added encapsulation for clarity.

commit 9c13c8918d63b46bcdafcc305bea4a60a118448f
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Mon Feb 28 22:40:33 2011 +0100

    Removed the function call parameter of *_set_gui_constraints ().

commit b86bde7d5480a2e4efa1949e53675df456ff9a8d
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Sun Feb 27 22:34:30 2011 +0100

    BGA: replaced tab for spaces.

commit 8ff67e85c25c7bf3935be73071591023d6d828c4
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Sun Feb 27 22:32:28 2011 +0100

    BGA: removed the function call parameter of bga_get_default_footprint_values().

commit ff9f3b5d7313466142ddc30c195aad9874691403
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Sun Feb 27 22:30:45 2011 +0100

    BGA: change the function return type of bga_create_packages_list().

commit 284b5a12168e5887e1e7891bd04636b9fb85c096
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Sun Feb 27 22:26:44 2011 +0100

    BGA: change the function return type of bga_create_element().

commit ed62e20a093d10dae81065c079db7b682a8933af
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Sun Feb 27 22:23:24 2011 +0100

    CAPA: removed the function call parameter of capa_get_default_footprint_values().

commit fec1e09a57a31c65b553ee20d13cf029e1f723fc
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Sun Feb 27 22:21:06 2011 +0100

    CAPA: change the function return type of capa_create_packages_list().

commit 68cf32a675c94e10f5161302a89e3f3547153eb5
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Sun Feb 27 22:18:12 2011 +0100

    CAPA: change the function return type of capa_create_element().

commit 7607ca6b443af63a0fd72acf89d5ac147af32722
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Sun Feb 27 22:14:51 2011 +0100

    CAPC: removed the function call parameter of capc_get_default_footprint_values().

commit 6ab8948e0d6d82315294c98b62b83a7ea8feaafa
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Sun Feb 27 22:13:03 2011 +0100

    CAPC: change the function return type of capc_create_packages_list().

commit 60363880be54cb533f1927a04e93922674e805ff
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Sun Feb 27 22:10:26 2011 +0100

    CAPC: change the function return type of capc_create_element().

commit c4eb401d61a4d408cc69830e8efcac1530514346
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Sun Feb 27 22:07:28 2011 +0100

    CAPM: removed the function call parameter of capm_get_default_footprint_values().

commit 3938db26918faf11b8708b63870f86ad400185fb
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Sun Feb 27 22:05:45 2011 +0100

    CAPM: change the function return type of capm_create_packages_list().

commit 8407aa973f2dd857884f63bb279cc0b7d1dfb568
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Sun Feb 27 22:03:31 2011 +0100

    CAPM: change the function return type of capm_create_element().

commit 2d3e3585a04a46e2c3ed1b78499b8ce582925371
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Sun Feb 27 21:49:49 2011 +0100

    CAPMP: removed the function call parameter of capmp_get_default_footprint_values().

commit c733b809fd5249fb0af662dfff7d544bbc885b66
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Sun Feb 27 21:48:05 2011 +0100

    CAPMP: change the function return type of capmp_create_packages_list().

commit 5cc1a8bdd1f362c0fc1b2a2fad19abfae657b96e
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Sun Feb 27 21:45:45 2011 +0100

    CAPMP: change the function return type of capmp_create_element().

commit c965ddcf2f7e15e60c0147b0696a4e2353867a79
Author: Bert Timmerman <bert.timmerman@xs4all.nl>
Date:   Sun Feb 27 21:42:48 2011 +0100

    CAPPR: removed the function call parameter of cappr_get_default_footprint_values().

commit 6f48132feb9d3f1b09c1b09a02dd05f748918fd6
Author: Bert Timmerman <bert.timmerman@xs4all.nl>

src/fpw.c
/*!
 * \file src/fpw.c
 *
 * \author Copyright (C) 2007, 2008, 2009, 2010 by Bert Timmerman <bert.timmerman@xs4all.nl>
 *
 * \brief fpw (FootPrintWizard) is a CLI program for the creation of
 * footprints.
 *
 * fpw (FootPrintWizard) is a program run from the Command Line Interface
 * (CLI) for the creation of footprints files.\n
 * These footprint files can be used by the pcb layout application
 * (see http://pcb.gpleda.org) to allow for the placement of parts on a
 * pcb layout.\n
 * \n
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.\n
 * \n
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.\n
 * \n
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.\n
 */


#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <string.h>
#include <stdio.h>
#include <getopt.h>
#include <gtk/gtk.h>
#include <stdlib.h>

/* Standard gettext macros. */
#ifdef ENABLE_NLS
#  include <libintl.h>
#  undef _
#  define _(String) dgettext (PACKAGE, String)
#  define Q_(String) g_strip_context ((String), gettext (String))
#  ifdef gettext_noop
#    define N_(String) gettext_noop (String)
#  else
#    define N_(String) (String)
#  endif
#else
#  define textdomain(String) (String)
#  define gettext(String) (String)
#  define dgettext(Domain,Message) (Message)
#  define dcgettext(Domain,Message,Type) (Message)
#  define bindtextdomain(Domain,Directory) (Domain)
#  define _(String) (String)
#  define Q_(String) g_strip_context ((String), (String))
#  define N_(String) (String)
        bindtextdomain (GETTEXT_PACKAGE, PACKAGE_LOCALE_DIR);
        bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
        textdomain (GETTEXT_PACKAGE);
#endif

#define GUI 0
/*!< Define whether this is a GTK User Interface (GUI) or not. */
#define CLI 1
/*!< Define whether this is a Command Line Interface (CLI) or not. */


#include "libfpw.c"
#include "packages.h"
#include "fpw.h"


gchar *program_name = NULL;
gchar *fpw_filename;


/*!
 * \brief Print the version of the footprintwizard to stderr.
 *
 * \return \c EXIT_SUCCESS.
 */
int
print_version ()
{
        fprintf (stderr, (_("\nfpw version %s\n")), VERSION);
        fprintf (stderr, (_("(C) 2007, 2008, 2009, 2010, 2011 by Bert Timmerman.\n")));
        fprintf (stderr, (_("This free software is released under the GPL v2 license;\n")));
        fprintf (stderr, (_("see the source for copying conditions.\n")));
        fprintf (stderr, (_("There is NO warranty; not even for MERCHANTABILITY\n")));
        fprintf (stderr, (_("or FITNESS FOR A PARTICULAR PURPOSE.\n\n")));
        return (EXIT_SUCCESS);
}


/*!
 * \brief Print the usage message for the footprintwizard to stderr.
 *
 * \return \c EXIT_SUCCESS.
 */
int
print_usage ()
{
        fprintf (stderr, (_("\nfpw usage and options:\n")));
        fprintf (stderr, (_("\t --help \n")));
        fprintf (stderr, (_("\t -? \n")));
        fprintf (stderr, (_("\t -h        : print this help message and exit.\n\n")));
        fprintf (stderr, (_("\t --verbose \n")));
        fprintf (stderr, (_("\t -v        : log messages, be verbose.\n\n")));
        fprintf (stderr, (_("\t --silent \n")));
        fprintf (stderr, (_("\t --quiet \n")));
        fprintf (stderr, (_("\t -q        : do not log messages, overrides --verbose.\n\n")));
        fprintf (stderr, (_("\t --version \n")));
        fprintf (stderr, (_("\t -V        : print the version information and exit.\n\n")));
        fprintf (stderr, (_("\t --format <fpw_footprintwizard filename> \n")));
        fprintf (stderr, (_("\t -f <fpw_footprintwizard filename>\n\n")));
        fprintf (stderr, (_("\t --output <footprint name> \n")));
        fprintf (stderr, (_("\t -o <footprint name>\n\n")));
        fprintf (stderr, (_("\t --debug \n")));
        fprintf (stderr, (_("\t -d        : turn on debugging output messages.\n\n")));
        return (EXIT_SUCCESS);
}


/*!
 * \brief The CLI for the footprintwizard.
 *
 * \return \c EXIT_SUCCESS.
 */
int
main
(
        int argc, /*!< : number of arguments on CLI */
        char **argv /*!< : array of argument variables */
)
{
        /* Determine how we are called today */
        program_name = argv[0];
        static const struct option opts[] =
        {
                {"debug", no_argument, NULL, 'd'},
                {"help", no_argument, NULL, 'h'},
                {"version", no_argument, NULL, 'V'},
                {"verbose", no_argument, NULL, 'v'},
                {"quiet", no_argument, NULL, 'q'},
                {"silent", no_argument, NULL, 'q'},
                {"format", required_argument, NULL, 'f'},
                {"output", required_argument, NULL, 'o'},
                {0, 0, 0, 0}
        };
        int optc;
        while ((optc = getopt_long (argc, argv, "dhVvqqf:o:", opts, NULL)) != -1)
        {
                switch (optc)
                {
                        case 'd':
                                debug = TRUE;
                                break;
                        case 'h':
                                print_usage ();
                                exit (EXIT_SUCCESS);
                        case 'V':
                                print_version ();
                                exit (EXIT_SUCCESS);
                        case 'v':
                                verbose = TRUE;
                                break;
                        case 'q':
                                silent = TRUE;
                                verbose = FALSE; /* Just to be sure. */
                                break;
                        case 'f':
                                fpw_filename = strdup (optarg);
                                if (debug)
                                        fprintf (stderr, "fpw filename = %s\n", fpw_filename);
                                break;
                        case 'o':
                                footprint_name = strdup (optarg);
                                if (debug)
                                        fprintf (stderr, "footprint name = %s\n", footprint_name);
                                break;
                        case '?':
                                print_usage ();
                                exit (EXIT_FAILURE);
                        default:
                                g_log ("", G_LOG_LEVEL_WARNING,
                                        _("unknown command line option encountered.\n"));
                                print_usage ();
                                exit (EXIT_FAILURE);
                }
        }
        if (optind < argc)
        {
                print_usage ();
                exit (EXIT_FAILURE);
        }
        /* Read variables from the fpw file */
        if (read_footprintwizard_file (fpw_filename))
        {
                if (verbose)
                {
                        g_log ("", G_LOG_LEVEL_INFO,
                                _("read footprintwizard file %s."),
                                fpw_filename);
                }
        }
        else
        {
                if ((verbose) || (!silent))
                {
                        g_log ("", G_LOG_LEVEL_ERROR,
                                _("could not load footprintwizard file %s."),
                                fpw_filename);
                }
                        exit (EXIT_FAILURE);
        }
        /* Check for a null pointer in footprint_name for this might cause a
         * segmentation fault or undefined behaviour. */
        if (!footprint_name)
        {
                if ((verbose) || (!silent))
                {
                        g_log ("", G_LOG_LEVEL_ERROR,
                                _("footprint name contains a null pointer."));
                }
                        exit (EXIT_FAILURE);
        }
        /* Check for an empty footprint_name string for this might cause a
         * segmentation fault or undefined behaviour. */
        if (!strcmp (footprint_name, ""))
        {
                if ((verbose) || (!silent))
                {
                        g_log ("", G_LOG_LEVEL_ERROR,
                                _("footprint name contains an empty string."));
                }
                exit (EXIT_FAILURE);
        }
        /* If the footprint_filename contains a valid footprintname, use it. */
        if (g_str_has_suffix (footprint_name, fp_suffix))
        {
                /* Footprintname has a .fp suffix, do nothing */
                footprint_filename = g_strdup (footprint_name);
        }
        else
        {
                /* Footprintname has no .fp suffix, add a .fp suffix */
                footprint_filename = g_strconcat (footprint_name, fp_suffix, NULL);
        }
        write_footprint ();
        if (verbose)
        {
                g_log ("", G_LOG_LEVEL_INFO,
                _("Footprint %s is written successful."),
                footprint_name);
        }
}


/* EOF */
